navigation/tracking/scripts/pdaf.py

```diff
@@ -0,0 +1,250 @@
+        o_predicted = np.matmul(self.C, self.state_pri)
+        diff = o - o_predicted
+        mah_dist = np.matmul(
+            diff.reshape(2, 1).T, np.matmul(np.linalg.inv(self.S), diff.reshape(2, 1))
+        )
+        return mah_dist
+
+    def filter_observations_outside_gate(self, o):
+
+        within_gate = []
+
+        for o_i in o:
+            mah_dist = self.compute_mah_dist(o_i)
+            if mah_dist < self.validation_gate_scaling_param**2:
+                within_gate.append(o_i)
+
+        self.o_within_gate_arr = np.array(within_gate)
+
+    def compute_probability_of_matching_observations(self):
+
+        score = np.ndarray(
+            (len(self.o_within_gate_arr),), dtype=float
+        )  # score for each observation based on distance from predicted position
+
+        self.p_match_arr = np.ndarray((len(self.o_within_gate_arr) + 1,), dtype=float)
+
+        if len(self.o_within_gate_arr) == 0:
+            self.p_match_arr[0] = 1.0
+        else:
+            self.p_match_arr[0] = self.p_no_match
+
+        for i, o_i in enumerate(self.o_within_gate_arr):
+
+            mah_distance = self.compute_mah_dist(o_i)
+            if (
+                mah_distance <= 0.1
+            ):  # In order to avoid infinte high weights. Choose an approriate threshold.
+                score[i] = 10
+            else:
+                score[i] = 1 / mah_distance
+
+        score_sum = np.sum(score)
+        for i in range(len(self.o_within_gate_arr)):
+            self.p_match_arr[i + 1] = (score[i] / score_sum) * (1 - self.p_no_match)
+
+    def compute_residual_vector(self):
+        self.residual_vector[0] = 0  # x
+        self.residual_vector[1] = 0  # y
+
+        for i in range(len(self.o_within_gate_arr)):
+            self.residual_vector[0] += self.p_match_arr[i + 1] * (
+                self.o_within_gate_arr[i][0] - self.state_pri[0]
+            )
+            self.residual_vector[1] += self.p_match_arr[i + 1] * (
+                self.o_within_gate_arr[i][1] - self.state_pri[1]
+            )
+
+    def compute_S(self):
+        C_P = np.matmul(self.C, self.P_pri)
+        self.S = np.matmul(C_P, self.C.T) + self.R
+
+    def compute_L(self):
+        P_CT = np.matmul(self.P_pri, self.C.T)
+        C_P_CT = np.matmul(self.C, P_CT)
+        self.L = np.matmul(P_CT, np.linalg.inv(C_P_CT + self.R))
+
+        # print("\n --- L ---\n")
+        # print(self.L)
+
+    def correct_state_vector(self):
+        self.state_post = self.state_pri + np.matmul(self.L, self.residual_vector)
+
+    def correct_P(self):
+        temp1 = np.ndarray((2, 2), dtype=float)
+        for i, o_i in enumerate(self.o_within_gate_arr):
+            ny_ak = o_i - np.matmul(self.C, self.state_pri)
+            temp1 += self.p_match_arr[i + 1] * np.matmul(ny_ak.reshape(2,1), ny_ak.reshape(2,1).T)
+        temp2 = temp1 - np.matmul(self.residual_vector.reshape(2,1), self.residual_vector.reshape(2,1).T)
```

Comment:
This is a bit hard to read, but could be improved by renaming temp1 and temp2, plus the @ notation